youtube_dl/extractor/torrins.py

```diff
@@ -0,0 +1,136 @@
+            error_data = error.get('data')
+            if error_data:
+                error_str += ' - %s' % error_data.get('formErrors')
+            raise ExtractorError(error_str, expected=True)
+
+    def _download_webpage(self, *args, **kwargs):
+        kwargs.setdefault('headers', {})['User-Agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_5) AppleWebKit/603.2.4 (KHTML, like Gecko) Version/10.1.1 Safari/603.2.4'
+        return super(TorrinsIE, self)._download_webpage(
+            *args, **compat_kwargs(kwargs))
+
+    def _real_initialize(self):
+        self._login()
+
+    def _login(self):
+        (username, password) = self._get_login_info()
+        if username is None:
+            return
+
+        self._download_webpage(
+            self._LOGIN_URL, None, 'Downloading login popup')
+
+        def is_logged(reason):
+            webpage = self._download_webpage(self._ORIGIN_URL, None, reason)
+
+            return any(re.search(p, webpage) for p in (
+                r'id=["\'](?:bt-logout)',
+                r'>Logout<'))
+
+        # already logged in
+        if is_logged('Checking if already logged in'):
+            return
+
+        login_form = {
+            'email': username,
+            'password': password,
+        }
+
+        response = self._download_webpage(
+            self._LOGIN_URL, None, 'Logging in',
+            data=urlencode_postdata(login_form),
+            headers={
+                'Referer': self._ORIGIN_URL,
+                'Origin': self._ORIGIN_URL,
+            })
+
+        if not is_logged('Post login check'):
+            error = self._html_search_regex(
+                r'(?s)<div[^>]+class="form-errors[^"]*">(.+?)</div>',
+                response, 'error message', default=None)
+            if error:
+                raise ExtractorError('Unable to login: %s' % error, expected=True)
+            raise ExtractorError('Unable to log in')
+
+    def _real_extract(self, url):
+        course_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, course_id)
+
+        title = self._og_search_title(webpage)
+
+        video_json = self._html_search_regex(r"<div id=\"video\" idata='(.+?)'", webpage, 'video formats')
+
+        video_json = self._parse_json(video_json, course_id)
+
+        title = video_json.get('title')
+        video_id = video_json.get('id')
+
+        formats = [
+            {
+                'url': video_json.get('low'),
+                'format_id': 'low',
+                'height': 240,
+                'ext': 'mp4'
+            },
+            {
+                'url': video_json.get('medium'),
+                'format_id': 'medium',
+                'height': 360,
+                'ext': 'mp4'
+            },
+            {
+                'url': video_json.get('high'),
+                'format_id': 'high',
+                'height': 480,
+                'ext': 'mp4'
+            }
+        ]
```

Comment:
Code duplication. `url` must not be `None`.

Reply:
Fixed.

Reply:
Nothing changed.

Reply:
Well, `url` cannot be none, I am not doing `'url': video_json.get('low')`.

Reply:
Now you break extraction if any of these keys is missing.

Reply:
From documentation:

> **Example**
> Say meta from the previous example has a title and you are about to extract it. Since title is a mandatory meta field you should end up with something like:
> 
> `title = meta['title']`
>
> If title disappears from meta in future due to some changes on the hoster's side the extraction would fail since title is mandatory. That's expected.

Isn't that expected?

Reply:
So what? How title applies here?

Reply:
I see what you mean here. I will make one more change.